### PR TITLE
Bug 62861

### PR DESCRIPTION
**Pull request by shaqianqian**
## Description
Add a button in thread group to control the clear of cache and cookies

## Motivation and Context
To simplify the control of clearing caches and cookies

## How Has This Been Tested?
ThreadGroup

### CODE REVIEW

File: test/src/org/apache/jmeter/control/TestThreadGroup.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        loop.setLoops(1);
+        loop.setContinueForever(false);
+        hashTree.add(loop);
+        hashTree.add(loop, transactionController);
+        threadGroup.setNumThreads(1);
+
+    }
+
+    @Test
+    public void testThreadLoaclForSameUserOnIternation() {
```
Comment (ubikloadpack) [inline test/src/org/apache/jmeter/control/TestThreadGroup.java]: Typo in name

File: src/core/org/apache/jmeter/threads/ThreadGroup.java [hunk 2]
```diff
@@ -69,7 +69,9 @@
 
     /** Scheduler start delay, overrides start time */
     public static final String DELAY = "ThreadGroup.delay";
-
+    
+    /** The same user or different users */
+    public static final String ISSAMEUSER = "ThreadGroup.sameuser";
```
Comment (ubikloadpack) [inline src/core/org/apache/jmeter/threads/ThreadGroup.java]: Rename to IS_SAME_USER

File: src/core/org/apache/jmeter/threads/ThreadGroup.java [hunk 3]
```diff
@@ -175,6 +177,26 @@ public void setRampUp(int rampUp) {
...
+    public void setIsSameUser(boolean isSameUser) {
+        setProperty(new BooleanProperty(ISSAMEUSER, isSameUser));
+    }
+
+    /**
+     * Get kind of user.
+     *
+     * @return the kind of user.
+     */
+    public boolean IsSameUser() {
```
Comment (ubikloadpack) [inline src/core/org/apache/jmeter/threads/ThreadGroup.java]: functions must start with lower case

File: src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java [hunk 4]
```diff
@@ -252,4 +269,17 @@ private void init() { // WARNING: called from ctor so must not be overridden (i.
         intgrationPanel.add(mainPanel);
         add(intgrationPanel, BorderLayout.CENTER);
     }
+   
+   private JPanel createUserOptionsPanel(){
+       ButtonGroup group = new ButtonGroup();
+       sameUserBox = new JRadioButton("Same user on each iteration"); //$NON-NLS-1$
```
Comment (ubikloadpack) [inline src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java]: Externalize labels into resource

File: src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java [hunk 5]
```diff
@@ -252,4 +269,17 @@ private void init() { // WARNING: called from ctor so must not be overridden (i.
...
+       group.add(sameUserBox);
+       sameUserBox.setSelected(true);
+       differentUserBox = new JRadioButton("Different User on Each Iteration"); //$NON-NLS-1$
```
Comment (ubikloadpack) [inline src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java]: Externalize labels into resource

File: src/protocol/http/org/apache/jmeter/protocol/http/gui/CacheManagerGui.java [hunk 6]
```diff
@@ -102,12 +110,19 @@ private void init() {
         setBorder(makeBorder());
 
         clearEachIteration = new JCheckBox(JMeterUtils.getResString("clear_cache_per_iter"), false); // $NON-NLS-1$
+        
+        controlledByThreadGroup = 
+                new JCheckBox("Use Thread Group configuration to control cache clearing", false); //$NON-NLS-1$
```
Comment (ubikloadpack) [inline src/protocol/http/org/apache/jmeter/protocol/http/gui/CacheManagerGui.java]: Externalize labels into resource

File: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java [hunk 7]
```diff
@@ -1727,6 +1736,7 @@ private void saveConnectionCookies(HttpResponse method, URL u, CookieManager coo
             Header[] hdrs = method.getHeaders(HTTPConstants.HEADER_SET_COOKIE);
             for (Header hdr : hdrs) {
                 cookieManager.addCookieFromHeader(hdr.getValue(),u);
+                log.info("The value of saveConnectionCookies are {}",hdr.getValue());
```
Comment (ubikloadpack) [inline src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java]: Why this log ? Make it more clear if needed

File: src/core/org/apache/jmeter/threads/JMeterThread.java [hunk 8]
```diff
@@ -238,9 +248,10 @@ private void startScheduler() {
     public void setThreadName(String threadName) {
         this.threadName = threadName;
     }
-
+    public static final ThreadLocal <Boolean>threadLocal4SameUser = new ThreadLocal<Boolean>() ;
```
Comment (ubikloadpack) [inline src/core/org/apache/jmeter/threads/JMeterThread.java]: Request was to make this field not Thread Local but a private JMeter Variable similar to the ones prefixed by __jmeter.....

**Review by ubikloadpack — CHANGES_REQUESTED** [review]

**Comment by shaqianqian** [comment]
Hello,
If don’t use threadlocal, how to share the value in the same thread? Thanks
for response.

UBIK LOAD PACK <notifications@github.com>于2019年6月2日 周日19:14写道：

**Comment by ubikloadpack** [comment]
The object JMeterVariables is specific to each Thread.

**Comment by shaqianqian** [comment]
Can you give me a example ?

UBIK LOAD PACK <notifications@github.com>于2019年6月2日 周日19:20写道：

**Comment by ubikloadpack** [comment]
Hi, Look at example in HTTPHC4Impl, or we'll discuss tomorrow.

**Comment by shaqianqian** [comment]
Sorry I dont have laptop at home, maybe we can talk about it tomorrow,
thanks for your help

UBIK LOAD PACK <notifications@github.com>于2019年6月2日 周日19:22写道：

**Comment by ubikloadpack** [comment]
No problem, review was for tomorrow :-)

File: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java [hunk 9]
```diff
@@ -1184,11 +1184,19 @@ protected void setupRequest(URL url, HttpRequestBase httpRequest, HTTPSampleResu
         } else {
             httpRequest.setHeader(HTTPConstants.HEADER_CONNECTION, HTTPConstants.CONNECTION_CLOSE);
         }
-    
-        setConnectionHeaders(httpRequest, url, getHeaderManager(), getCacheManager());
-    
+        JMeterVariables variables = JMeterContextService.getContext().getVariables();
```
Comment (ubikloadpack) [inline src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java:1187]: This should be in CacheManager#testIterationStart

File: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java [hunk 10]
```diff
@@ -1184,11 +1184,19 @@ protected void setupRequest(URL url, HttpRequestBase httpRequest, HTTPSampleResu
...
+        JMeterVariables variables = JMeterContextService.getContext().getVariables();       
+        if(variables.getObject(JMeterThread.IS_SAME_USER)!=null&&getCacheManager()!=null&&
+                !(boolean)variables.getObject(JMeterThread.IS_SAME_USER)&&getCacheManager().getControlledByThread()) {
+            getCacheManager().setClearEachIteration(true);
+            
+        }
+        setConnectionHeaders(httpRequest, url, getHeaderManager(), getCacheManager());      
+        if(variables.getObject(JMeterThread.IS_SAME_USER)!=null&&getCookieManager()!=null&&
+                !(boolean)variables.getObject(JMeterThread.IS_SAME_USER)&& getCookieManager().getControlledByThread()) {
+            getCookieManager().setClearEachIteration(true);
```
Comment (ubikloadpack) [inline src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java:1196]: This should be in CookieManager#testIterationStart

**Review by ubikloadpack — CHANGES_REQUESTED** [review]

**Comment by ubikloadpack** [comment]
PR is incomplete and needs rework, see mail I sent you.
Thanks

**Pull request closed without merging**